internal/generator/generator_util.go

```diff
@@ -0,0 +1,238 @@
+
+	// Extract one directive definition and attributes from extracted block
+	// { ngx_string({directive_name}),
+	//   {bitmask1|bitmask2|...},
+	//   ... },
+	// this regex extracts {directive_name} and {bitmask1|bitmask2|...}.
+	extractSingleDirective = "ngx_string\\(\"(.*?)\"\\).*?,(.*?),"
+
+	extractSingleLineComment = `//.*`
+
+	extractMultiLineComment = `/\*[\s\S]*?\*/`
+)
+
+type bitDef []string
+
+type supFileTmplStruct struct {
+	Directive2bitDefs map[string][]bitDef
+	MapVariableName   string
+	MatchFnName       string
+}
+
+var (
+	directivesDefBlockExtracter = regexp.MustCompile(extractDirectivesDefBlock)
+	singleDirectiveExtracter    = regexp.MustCompile(extractSingleDirective)
+	singleLineCommentExtracter  = regexp.MustCompile(extractSingleLineComment)
+	multiLineCommentExtracter   = regexp.MustCompile(extractMultiLineComment)
+)
+
+//go:embed tmpl/support_file.tmpl
+var supFileTmplStr string
+
+//nolint:gochecknoglobals
+var supFileTmpl = template.Must(template.New("supFile").
+	Funcs(template.FuncMap{"Join": strings.Join}).Parse(supFileTmplStr))
+
+//nolint:gochecknoglobals
+var ngxBitmaskNameToGo = map[string]string{
+	"NGX_MAIL_MAIN_CONF":   "ngxMailMainConf",
+	"NGX_STREAM_MAIN_CONF": "ngxStreamMainConf",
+	"NGX_CONF_TAKE1":       "ngxConfTake1",
+	"NGX_STREAM_UPS_CONF":  "ngxStreamUpsConf",
+	"NGX_HTTP_LIF_CONF":    "ngxHTTPLifConf",
+	"NGX_CONF_TAKE2":       "ngxConfTake2",
+	"NGX_HTTP_UPS_CONF":    "ngxHTTPUpsConf",
+	"NGX_CONF_TAKE23":      "ngxConfTake23",
+	"NGX_CONF_TAKE12":      "ngxConfTake12",
+	"NGX_HTTP_MAIN_CONF":   "ngxHTTPMainConf",
+	"NGX_HTTP_LMT_CONF":    "ngxHTTPLmtConf",
+	"NGX_CONF_TAKE1234":    "ngxConfTake1234",
+	"NGX_MAIL_SRV_CONF":    "ngxMailSrvConf",
+	"NGX_CONF_FLAG":        "ngxConfFlag",
+	"NGX_HTTP_SRV_CONF":    "ngxHTTPSrvConf",
+	"NGX_CONF_1MORE":       "ngxConf1More",
+	"NGX_ANY_CONF":         "ngxAnyConf",
+	"NGX_CONF_TAKE123":     "ngxConfTake123",
+	"NGX_MAIN_CONF":        "ngxMainConf",
+	"NGX_CONF_NOARGS":      "ngxConfNoArgs",
+	"NGX_CONF_2MORE":       "ngxConf2More",
+	"NGX_CONF_TAKE3":       "ngxConfTake3",
+	"NGX_HTTP_SIF_CONF":    "ngxHTTPSifConf",
+	"NGX_EVENT_CONF":       "ngxEventConf",
+	"NGX_CONF_BLOCK":       "ngxConfBlock",
+	"NGX_HTTP_LOC_CONF":    "ngxHTTPLocConf",
+	"NGX_STREAM_SRV_CONF":  "ngxStreamSrvConf",
+	"NGX_DIRECT_CONF":      "ngxDirectConf",
+	"NGX_CONF_TAKE13":      "ngxConfTake13",
+	"NGX_CONF_ANY":         "ngxConfAny",
+	"NGX_CONF_TAKE4":       "ngxConfTake4",
+}
+
+//nolint:gochecknoglobals
+var allDirectiveContexts = map[string]interface{}{
+	"ngxMainConf":       nil,
+	"ngxEventConf":      nil,
+	"ngxMailMainConf":   nil,
+	"ngxMailSrvConf":    nil,
+	"ngxStreamMainConf": nil,
+	"ngxStreamSrvConf":  nil,
+	"ngxStreamUpsConf":  nil,
+	"ngxHTTPMainConf":   nil,
+	"ngxHTTPSrvConf":    nil,
+	"ngxHTTPLocConf":    nil,
+	"ngxHTTPUpsConf":    nil,
+	"ngxHTTPSifConf":    nil,
+	"ngxHTTPLifConf":    nil,
+	"ngxHTTPLmtConf":    nil,
+	"ngxMgmtMainConf":   nil,
+}
```

Comment:
Have you come across any examples of dynamic modules creating their own contexts? Without spending too much time looking I've only found examples that work something like `mgmt` (`geoip2`, `otel`).

Reply:
> The better way to do this is with a map[string]struct{}.

Sure! Thanks for telling me this. Will fix in next commit

> I've also seen this as map[string]bool that I think is worse because there are 3 possible values: unset, true, or false.

I found this only in an existing test `TestAnalyze`. It seems nothing to do with this MR, maybe I can change it in another separate MR?

> Have you come across any examples of dynamic modules creating their own contexts?

No at all... The only new context I've met is `ngxMgmtMainConf`(and this is created by us, not in source code). I have a map from C style context to go style and if there is a new context not in map I will get an error. Till now I haven't met an error for it.